### DISCUSSION TITLE
Non-Generic Inner Functions?

### DISCUSSION THREAD

**Discussion by PatatasDelPapa** [trillium-rs/trillium#37] [Ideas]
This morning I've been shared [this article](https://www.possiblerust.com/pattern/non-generic-inner-functions) about what says in the title and looking at the code of some methods taking for example `impl Into<Body>` then doing the into and performing the rest of the function with the concrete Body like [this function ](https://docs.trillium.rs/src/trillium_http/conn.rs.html#211-220) it seems to me that this is a prime example to apply this pattern (Having the generic parameter and an inner function working with the concrete type after the conversion). 

The idea behind is explained in the article but this part capture it well enough
> Most of the function body isn’t generic, the generic is simply to be more flexible in the input type, which the function then converts into a single concrete type. So, you can create a separate function which takes the converted-to type, and then have the generic function only perform the conversion and then call the non-generic function. This way, the time spent generating code and the size of the generic code in the resulting binary are both reduced.

**Comment by jbr** [comment]
Is the motivation performance in this specific context? There's probably a lot of benchmark-driven perf work to be done on trillium in the near future, and I'll definitely keep this in mind as a possible optimization. In general, my focus is initially on adding feature support so that people can express complete applications, with perf being a later concern that only matters if there's sufficient adoption

**Reply by PatatasDelPapa** [reply]
This one motivation is reducing binary size by reducing code to that needs to be monophormized (is that the correct term?) but I agree that maybe it's too early to focus on this kind of optimizations.
